File: go/hilo.go
package main

import (
	"bufio"
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"strings"
	"time"
)

func hilo() {
	for {
		fmt.Println("\nWelcome to Hi-Lo! The rules are simple: you have 7 tries to guess a number 1-100.")
		fmt.Println("If your guess is too low, I'll say 'Too low'; if your guess is too high, I'll say 'Too high'.")
		fmt.Println("If you guess correctly, you win!")

		rand.Seed(time.Now().UTC().UnixNano())
		reader := bufio.NewReader(os.Stdin)

		answer := rand.Intn(100) + 1

		guess := 0
		var err error
		for attempt := 1; attempt <= 7; attempt++ {
			for {
				fmt.Printf("\nGuess #%v (1-100): ", attempt)
				guessStr, _ := reader.ReadString('\n')
				guess, err = strconv.Atoi(strings.Trim(guessStr, "\n"))
				if guessStr != "\n" && err == nil && guess >= 1 && guess <= 100 {
					break
				}
			}

			if guess > answer {
				fmt.Println("Too high.")
			} else if guess < answer {
				fmt.Println("Too low.")
			} else {
				break
			}
		}

		if guess == answer {
			fmt.Println("You win!")
		} else {
			fmt.Printf("You lost. The correct number was %v.\n", answer)
		}

		fmt.Print("Would you like to play again? (y/n): ")
		choice, _ := reader.ReadString('\n')
		if strings.Trim(choice, "\n") != "y" {
			print("Thanks for playing!")
			break
		}
	}

}

func main() {
	hilo()
}

File: c/hilo.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void hilo() {
    while (true) {
        printf(
            "\nWelcome to Hi-Lo! The rules are simple: you have 7 tries to "
            "guess a number 1-100.\nIf your guess is too low, I'll say \'Too "
            "low\'; if your guess is too high, I'll say \'Too high\'.\nIf you "
            "guess correctly, you win!\n\n");

        int answer = rand() % 100 + 1;

        int guess = 0;
        for (int attempt = 1; attempt <= 7; ++attempt) {
            while (true) {
                printf("Guess #%d (1-100): ", attempt);

                int input = 0;
                scanf("%d", &input);

                guess = input;
                if (guess >= 1 && guess <= 100) {
                    break;
                }
                while (getchar() != '\n')
                    ;
            }

            if (guess < answer) {
                printf("Too low.\n");
            } else if (guess > answer) {
                printf("Too high.\n");
            } else {
                printf("You win!\n");
                break;
            }
        }

        if (guess != answer) {
            printf("You lost. The correct number was %d.\n", answer);
        }

        printf("Would you like to play again? (y/n): ");
        char choice[] = "n";
        scanf("%s", choice);
        if (choice[0] != 'y') {
            printf("Thanks for playing!\n");
            break;
        }
    }
}

int main() {
    srand(time(NULL));

    hilo();

    return 0;
}
